### QDNAseq: Quantitative DNA sequencing for chromosomal aberrations

This repository contains source code for the R/Bioconductor package
[QDNAseq](http://bioconductor.org/packages/release/bioc/html/QDNAseq.html).
For a description, please see
[this article](http://genome.cshlp.org/content/24/12/2022.long),
and for usage, the vignette from Bioconductor
[development version](http://bioconductor.org/packages/devel/bioc/vignettes/QDNAseq/inst/doc/QDNAseq.pdf)
or the
[latest release](http://bioconductor.org/packages/release/bioc/vignettes/QDNAseq/inst/doc/QDNAseq.pdf)

#### Citing QDNAseq

To cite QDNAseq, please use:  
Scheinin I, Sie D, Bengtsson H, van de Wiel MA, Olshen AB, van Thuijl HF, van
Essen HF, Eijk PP, Rustenburg F, Meijer GA, Reijneveld JC, Wesseling P, Pinkel
D, Albertson DG and Ylstra B (2014) **DNA copy number analysis of fresh and
formalin-fixed specimens by shallow whole-genome sequencing with identification
and exclusion of problematic regions in the genome assembly.** *Genome
Research* **24**: 2022-2032

#### Software quality

* R CMD check status:
  - Bioconductor development version: [![BUILD/CHECK Report](http://bioconductor.org/shields/build/devel/bioc/QDNAseq.svg)](http://master.bioconductor.org/checkResults/devel/bioc-LATEST/QDNAseq/)
  - Bioconductor release version: [![BUILD/CHECK Report](http://bioconductor.org/shields/build/release/bioc/QDNAseq.svg)](http://master.bioconductor.org/checkResults/release/bioc-LATEST/QDNAseq/)
  - Travis CI: [![Build Status](https://travis-ci.org/ccagc/QDNAseq.svg?branch=master)](https://travis-ci.org/ccagc/QDNAseq)
* Test coverage status:
  - Coveralls CI: [![Coverage Status](https://coveralls.io/repos/ccagc/QDNAseq/badge.png?branch=master)](https://coveralls.io/r/ccagc/QDNAseq?branch=master)
